### PR TITLE
Shadow culling outside view frustum

### PR DESCRIPTION
**Pull request by MateuszWasilewski**
## What does this PR do?

As of right now, when there are many lights on the scene, then usually most of them are outside the Camera frustum. Each light has a max range that it can cast shadow to. 
But current implementation always goes through object culling, shadowmap rendering etc. for all lights. This happens even if all shadows possibly casted from the light source are outside camera view.

This PR aims to improve performance by skipping unnecessary work for all shadows that won't be visible. 
This is achieved by culling light on the CPU to limit the amount of lights that processing is done for. 
Used intersection test: sphere (center=lightPosition, radius=maxRange) with view frustum of every persistent View.

There are other scenarios when culling shadows is also possible.
I hope that minimizing light impact on performance does not end here.
Potential further improvements:
- Free memory on the GPU for culled shadows
- Take culling planes and big objects into consideration when culling shadows
- etc.

CVAR `r_cullShadowmapOutsideViewFrustum` was added make this feature opt out if needed.

## How was this PR tested?
Manually by verifying on different scenes if shadows look the same with CVAR `r_cullShadowmapOutsideViewFrustum` set to 0/1.

After applying the PR to scene with 900 lights, avg 100 meshes for light. 
CPU usage from CullingScene drastically decreased.

## Visualization
Minimal example with single light.

Improvements because of this PR:
- Decrease in total draw item count 17 -> 15, scales with each mesh / light.

Light in view (culling enabled / disabled no difference)
Screenshot from 2024-03-26 16-03-07

Light outside view (culling enabled)
Screenshot from 2024-03-26 16-09-07

Light outside view (culling disabled)
Screenshot from 2024-03-26 16-09-01

### CODE REVIEW

**Comment by michalpelka** [comment]
Thanks, @MateuszWasilewski for this PR! I've added the proper sig and set the priority level. Please feel free to adjust accordingly. 
It is an important feature for large indoor environments (tunnels, storage hall).

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Public/FeatureProcessor.h [hunk 1]
```diff
@@ -48,9 +48,10 @@ namespace AZ
             friend class Scene;
         public:
 
-            // [GFX TODO]: now these structure are empty, but we will clean up them later when we are sure we won't have any members in them.
+        public:
```
Comment (moudgils) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Public/FeatureProcessor.h:51]: Remove this line

File: Gems/Atom/Feature/Common/Code/Source/Shadows/ProjectedShadowFeatureProcessor.cpp [hunk 2]
```diff
@@ -631,23 +664,41 @@ namespace AZ::Render
...
                 for (ShadowProperty& shadowProperty : shadowProperties)
                 {
                     uint16_t shadowIndex = shadowProperty.m_shadowId.GetIndex();
                     const FilterParameter& filterData = m_shadowData.GetElement<FilterParamIndex>(shadowIndex);
-                    if (filterData.m_shadowmapSize == aznumeric_cast<uint32_t>(ShadowmapSize::None))
+                    bool needsRender = filterData.m_shadowmapSize != aznumeric_cast<uint32_t>(ShadowmapSize::None);
+                    if (!needsRender)
+                    {
+                        continue;
+                    }
```
Comment (moudgils) [inline Gems/Atom/Feature/Common/Code/Source/Shadows/ProjectedShadowFeatureProcessor.cpp:693]: Any reason to change the existing code as we are not using "needsRender" below?

**Review by moudgils — APPROVED** [review]
Minor comments.

**Comment by MateuszWasilewski** [comment]
After further testing i had to remove enabling / disabling shadow passes. It does not impact performance improvements directly, just UI visualization.
Enabling / disabling passes does not work as expected with low frame rates < 20fps

**Review by lsemp3d — APPROVED** [review]

**Pull request merged**